src/envoy/http/authn/authenticator_base.cc

```diff
@@ -35,33 +35,28 @@ AuthenticatorBase::~AuthenticatorBase() {}
 
 void AuthenticatorBase::done(bool success) const { done_callback_(success); }
 
-void AuthenticatorBase::validateX509(
-    const iaapi::MutualTls& mtls,
-    const AuthenticatorBase::MethodDoneCallback& done_callback) const {
+bool AuthenticatorBase::validateX509(
+  const iaapi::MutualTls& mtls, Payload& payload) const {
   const Network::Connection* connection = filter_context_.connection();
   if (connection == nullptr || connection->ssl() == nullptr) {
     // Not a TLS connection
-    done_callback(nullptr, false);
-    return;
+    return false;
   }
 
-  Payload payload;
   bool has_user =
       connection->ssl()->peerCertificatePresented() &&
       Utils::GetSourceUser(connection, payload.mutable_x509()->mutable_user());
 
   if (!has_user && !mtls.allow_tls()) {
```

Comment:
not exactly the same ? ex, A=F, B=F, !A&&!B is T, A||B is F

Reply:
Right, but you return the opposite, do you?

Anyway, the key point is return the value of the expression itself, instead of putting if/else for that.

Reply:
done